### ISSUE TITLE
Python3: Nuitka leaks references when class variables are accessed on class level

### ISSUE THREAD

**Issue by EnzoRondo** [kayhayen/Nuitka#107]
```
nuitka version 0.5.33rc7 (factory branch)
Microsoft Windows Server 2012 Standard
Python 3.6.6 (v3.6.6:4cf1f54eb7, Jun 27 2018, 03:37:03) [MSC v.1900 64 bit (AMD64)] on win32
```

I am using [grab.spider](https://github.com/lorien/grab) framework for my work, my script is simple web crawler which download web pages and find specified signs on that, grab framework uses threading, pycurl, lxml, and other libraries inside.

Now I have serious problem. When I am working with usual python bot runs at 80 threads and starts nearly from 100mb of RAM usage, then it grows to 500mb step by step, and then releases it back to ~200mb, all time I see that script takes up to 500mb and then moves down to 200mb.

Python version do this memory releases proper way, but when I am using nuitka I see only growth of memory and nothing more, nuitka not able to release memory. I tried to wait for a long time, not works, memory just goes to up and never go down.

I have tried to make simple tests to reproduce this situation, like that:
```python
import time
import psutil
import os
import humanfriendly


def mem_usage():
    process = psutil.Process(os.getpid())
    res = process.memory_info().rss
    return res


test = {}


def doit():
    global test
    test['abc'] = "123" * 200 * 200 * 200 * 40
    print(f"used top: {humanfriendly.format_size(mem_usage())}")
    time.sleep(1)


print(f"used before: {humanfriendly.format_size(mem_usage())}")
doit()
test = None
# time.sleep(1)
print(f"used after: {humanfriendly.format_size(mem_usage())}")
```

but here is no problem, nuitka releases memory in this simple example. My problem located somewhere deeper, and I have lack of skills what to do next. It would be great to provice you simple example, but grab library too complex thing which too difficult for me to analyze that.

Dear @kayhayen, could you give me any advices or recommendations how it possible to handle that?

My current plan: when bot finishing work need insert pdb.set_trace() and try to call gc.collect() and see reaction, maybe memory will released

**Comment by kayhayen** [comment]
I just checked the "tp_visit" slots of compiled types. Some time ago they were not up to the task, but nowadays they should be.

Call gc.collect() then do gc.get_objects() to get a list of all objects in the Python interpreter,do not store it anywhere however, make a del on the local variable you store it too.

Then check types of objects, and count them. And see what is growing. There is also gc.get_referrers and gc.get_referrents for given objects, to check what is holding objects in memory that you think should not.

Options:

a) Ultimately a cycle may not be detected. You might find that by the number of different kinds of objects that use another not being released. A minimal example of that would be great.

b) There is pretty extensive tests in Nuitka that show that references are not leaked. Maybe it's not detecting one specific case though, then this should be isolated and fixed.

BTW: Using pdb might not be working with compiled code.

**Comment by kayhayen** [comment]
I have recently discovered that for Python3 classes there is a leak and fixed it for the factory branch. Could well have to do with your issue, if class body code is run frequently for your program. It's going to be in a pre-release next week.

**Comment by EnzoRondo** [comment]
> Could well have to do with your issue, if class body code is run frequently for your program.
In my case classes used are extremely frequently and widely. I will test new version ASAP. Please let me know if something even potentially related to memory leaks will fixed.

About your idea with GC I still can't test it, too difficult for my current level.

**Comment by EnzoRondo** [comment]
> Could well have to do with your issue, if class body code is run frequently for your program.

In my case classes used are extremely frequently and widely. I will test new version ASAP. Please let me know if something even potentially related to memory leaks will fixed.

About your idea with GC I still can't test it, too difficult for my current level.

**Comment by EnzoRondo** [comment]
Thanks a lot dear friend, now works perfect! No leaks more, memory releases proper way as expected. You are awesome! Thank you for doing great things! I'm very happy from that.

**Comment by kayhayen** [comment]
Thanks, so we are lucky to have the coincidence of my finding this during my C bool work on Nuitka, be warned that i am making factory broken often. The release that fixes it for you will be 0.6 or the prelease I will be making of it early next week. Due to the complex changes in code generation I want to avoid doing a backport of the fix, so stable Nuitka will be affected for some longer. But "develop" is usually pretty stable for Nuitka.

**Comment by kayhayen** [comment]
This is in the current pre-release which is become a release any day now, closing therefore.
